reasoner/planner/AnswerCountEstimator.java

```diff
@@ -0,0 +1,505 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.vaticle.typedb.common.collection.Collections.list;
+import static com.vaticle.typedb.common.collection.Collections.set;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class AnswerCountEstimator {
+    private final LogicManager logicMgr;
+    private final Map<ResolvableConjunction, ConjunctionAnswerCountEstimator> estimators;
+
+    // Cycle handling
+    private final ArrayList<ResolvableConjunction> initializationStack;
+    private final Set<ResolvableConjunction> cycleHeads;
+    private final Set<ResolvableConjunction> toReset;
+
+    public AnswerCountEstimator(LogicManager logicMgr) {
+        this.logicMgr = logicMgr;
+        this.estimators = new HashMap<>();
+        this.initializationStack = new ArrayList<>();
+        this.cycleHeads = new HashSet<>();
+        this.toReset = new HashSet<>();
+    }
+
+    public long estimateAllAnswers(ResolvableConjunction conjunction) {
+        registerConjunctionAndInitialize(conjunction);
+        return estimators.get(conjunction).estimateAllAnswers();
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variableFilter) {
+        return estimateAnswers(conjunction, variableFilter, logicMgr.compile(conjunction));
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variableFilter, Set<Resolvable<?>> includedResolvables) {
+        registerConjunctionAndInitialize(conjunction);
+        return estimators.get(conjunction).estimateAnswers(variableFilter, includedResolvables);
+    }
+
+    private void registerConjunctionAndInitialize(ResolvableConjunction conjunction) {
+        if (!estimators.containsKey(conjunction)) {
+            estimators.put(conjunction, new ConjunctionAnswerCountEstimator(this, conjunction));
+        }
+
+        if (estimators.get(conjunction).mayInitialize()) {
+            initializationStack.add(conjunction);
+            estimators.get(conjunction).initialize();
+            assert initializationStack.get(initializationStack.size() - 1) == conjunction;
+            initializationStack.remove(initializationStack.size() - 1);
+
+            // In cyclic paths
+            if (cycleHeads.size() == 1 && cycleHeads.contains(conjunction)) {
+                // We can reset and recompute all!
+                for (ResolvableConjunction conjunctionToReset : toReset) {
+                    estimators.get(conjunctionToReset).reset();
+                }
+                estimators.get(conjunction).initialize(); // no loop needed. This is the maximal cycle.
+            }
+        }
+    }
+
+    private void reportInitializationCycle(ResolvableConjunction conjunction) {
+        cycleHeads.add(conjunction);
+        for (int i = initializationStack.size() - 1; initializationStack.get(i) != conjunction; i--) {
+            toReset.add(initializationStack.get(i));
+        }
```

Comment:
Mark all nodes in the cycle (except the cycle-head)  one as needing to be reset.